File: cmux-tui/bindings/conformance/raw/adapters/typescript/adapter.mjs
#!/usr/bin/env node
// Protocol-10 conformance adapter for the public TypeScript/Node SDK.

import {
  CmuxAuthorityError,
  CmuxClient,
  CmuxCommandError,
  CmuxProtocolError,
  CmuxTimeoutError,
  COMMAND_METADATA,
  EVENT_METADATA,
  UnixSocketTransport,
} from "../../../../typescript/dist/src/raw/index.js";
import { Buffer } from "node:buffer";

const UINT64_KEYS = new Set([
  "client",
  "index",
  "offset",
  "pane",
  "pane_revision",
  "projection_revision",
  "request",
  "screen",
  "seq",
  "surface",
  "terminal_revision",
  "timeout_ms",
  "workspace",
  "workspace_revision",
]);

function normalize(value, key = undefined) {
  if (typeof value === "bigint") return value.toString();
  if (value instanceof Uint8Array) return Buffer.from(value).toString("base64");
  if (Array.isArray(value)) return value.map((item) => normalize(item));
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value)
        .filter(([, item]) => item !== undefined)
        .map(([itemKey, item]) => [itemKey, normalize(item, itemKey)]),
    );
  }
  if (
    typeof value === "number"
    && Number.isInteger(value)
    && (UINT64_KEYS.has(key) || key?.endsWith("_revision"))
  ) {
    return String(value);
  }
  return value;
}

function classify(error) {
  const text = String(error?.message ?? error).toLowerCase();
  if (error instanceof CmuxTimeoutError || text.includes("timed out") || text.includes("did not produce")) {
    return "timeout";
  }
  if (text.includes("exceed") || text.includes("limit") || text.includes("buffer overflow")) {
    return "limit";
  }
  if (error instanceof CmuxCommandError) return "command";
  if (
    error instanceof CmuxProtocolError
    || text.includes("utf-8")
    || text.includes("json")
    || text.includes("decode")
  ) {
    return "decode";
  }
  return "transport";
}

function metadata() {
  return {
    commands: Object.entries(COMMAND_METADATA).map(([name, item]) => ({
      name,
      authority: item.authority,
      stream: item.stream?.kind ?? null,
    })),
    events: Object.entries(EVENT_METADATA).map(([name, item]) => ({
      name,
      streams: [...item.streams],
    })),
  };
}

function makeClient(request, enableProviderAuthority = false) {
  const maxInboundMessageBytes = Number(request.max_frame_bytes ?? 16 * 1024 * 1024);
  const transportOptions = { maxInboundMessageBytes };
  const makeTransport = () => new UnixSocketTransport(request.socket_path, transportOptions);
  const authorities = request.authority === "local-admin"
    ? ["frontend", "local-admin"]
    : undefined;
  return new CmuxClient({
    socketPath: request.socket_path,
    timeoutMs: Number(request.timeout_ms ?? 1000),
    maxBufferedEvents: Number(request.max_buffered_events ?? 256),
    authorities,
    enableProviderAuthority,
    transport: makeTransport(),
    streamTransportFactory: makeTransport,
  });
}

async function identify(request) {
  const client = makeClient(request);
  try {
    const value = await client.identify();
    return {
      app: value.app,
      protocol: value.protocol,
      workspace_revision: value.workspace_revision.toString(),
      terminal_revision: value.terminal_revision.toString(),
    };
  } finally {
    await client.close();
  }
}

async function nullableLiteral(request) {
  const client = makeClient(request);
  try {
    const value = await client.createTerminal({ key: "workspace-key" });
    return { lifecycle: value.lifecycle };
  } finally {
    await client.close();
  }
}

async function optionalNonNullResponse(request) {
  const client = makeClient(request);
  try {
    const value = await client.identify();
    return { present: value.capabilities !== undefined };
  } finally {
    await client.close();
  }
}

async function optionalNullableRequest(request) {
  const client = makeClient(request);
  const presence = String(request.presence);
  try {
    switch (presence) {
      case "omitted":
        await client.request("set-client-info", {});
        break;
      case "null":
        await client.request("set-client-info", { name: null });
        break;
      case "value":
        await client.request("set-client-info", { name: "conformance-client" });
        break;
      default:
        throw new Error(`unknown presence ${presence}`);
    }
    return { presence };
  } finally {
    await client.close();
  }
}

async function openStream(client, request) {
  switch (request.stream) {
    case "subscribe-coarse":
      return client.subscribe();
    case "subscribe-deltas":
      return client.subscribe({ treeEvents: "deltas" });
    case "attach-byte":
      return client.attachSurface(BigInt(request.surface ?? "7"), { mode: "bytes" });
    case "attach-render":
      return client.attachSurface(BigInt(request.surface ?? "7"), { mode: "render" });
    case "attach-browser":
      return client.attachSurface(BigInt(request.surface ?? "7"), { mode: "bytes" });
    default:
      throw new Error(`unknown stream ${String(request.stream)}`);
  }
}

function normalizeEvent(event) {
  if (!(event.event in EVENT_METADATA)) {
    return { event: event.event, unknown: true, raw: normalize(event) };
  }
  return normalize(event);
}

async function stream(request) {
  const client = makeClient(request);
  const events = [];
  let terminal = false;
  let opened;
  try {
    opened = await openStream(client, request);
    for (let index = 0; index < Number(request.events ?? 1); index += 1) {
      try {
        const event = await opened.next(Number(request.timeout_ms ?? 1000));
        events.push(normalizeEvent(event));
        if (event.event === "overflow" || event.event === "detached") terminal = true;
      } catch (error) {
        if (terminal || String(error?.message ?? error).includes("closed")) {
          terminal = true;
          break;
        }
        throw error;
      }
    }
    return { events, terminal };
  } finally {
    opened?.close();
    await client.close();
  }
}

async function requiredNullableEvent(request) {
  const client = makeClient(request);
  let opened;
  try {
    opened = await client.subscribe();
    const event = await opened.next(Number(request.timeout_ms ?? 1000));
    if (event.event !== "client-changed") {
      throw new CmuxProtocolError(`expected client-changed event, got ${event.event}`);
    }
    return { name: event.name };
  } finally {
    opened?.close();
    await client.close();
  }
}

async function optionalNonNullEvent(request) {
  const client = makeClient(request);
  let opened;
  try {
    opened = await openStream(client, request);
    const event = await opened.next(Number(request.timeout_ms ?? 1000));
    if (event.event !== "output") {
      throw new CmuxProtocolError(`expected output event, got ${event.event}`);
    }
    return { present: event.colors !== undefined };
  } finally {
    opened?.close();
    await client.close();
  }
}

async function closePendingStream(request) {
  const client = makeClient(request);
  let opened;
  try {
    opened = await openStream(client, request);
    const pending = opened.next(Number(request.deadline_ms ?? 1000) * 4)
      .then(() => true, () => true);
    await new Promise((resolve) => setTimeout(resolve, Number(request.close_after_ms ?? 50)));
    opened.close();
    const unblocked = await Promise.race([
      pending,
      new Promise((resolve) => setTimeout(() => resolve(false), Number(request.deadline_ms ?? 1000))),
    ]);
    return { unblocked };
  } finally {
    opened?.close();
    await client.close();
  }
}

async function authority(request) {
  const client = makeClient(request, request.authority === "provider-authority");
  let command;
  try {
    switch (request.authority) {
      case "control":
        await client.ping();
        command = "ping";
        break;
      case "frontend":
        await client.browserBack({ surface: 7n });
        command = "browser-back";
        break;
      case "local-admin":
        await client.pairingResponse({ request: 1n, approve: false });
        command = "pairing-response";
        break;
      case "provider-authority":
        await client.markWorkspacesProviderManaged({ authority: "conformance-authority" });
        command = "mark-workspaces-provider-managed";
        break;
      default:
        throw new Error(`unknown authority ${String(request.authority)}`);
    }
  } finally {
    await client.close();
  }
  return { command };
}

async function authorityDenied(request) {
  const client = makeClient(request);
  try {
    await client.markWorkspacesProviderManaged({ authority: "conformance-authority" });
  } catch (error) {
    if (error instanceof CmuxAuthorityError) return { denied: true };
    throw error;
  } finally {
    await client.close();
  }
  throw new Error("default client allowed provider-authority command");
}

function findSurface(tree, surface) {
  for (const workspace of tree.workspaces) {
    for (const screen of workspace.screens) {
      for (const pane of screen.panes) {
        if (!("tabs" in pane)) continue;
        const tab = pane.tabs.find((item) => item.surface === surface);
        if (tab) return { workspace, screen, pane, tab };
      }
    }
  }
  return undefined;
}

async function realFlow(request) {
  const marker = String(request.marker ?? "cmux-sdk-conformance-marker");
  const workspaceName = String(request.workspace_name ?? "sdk-conformance-workspace");
  const renamedName = String(request.renamed_name ?? "sdk-conformance-renamed");
  const client = makeClient(request);
  let opened;
  let workspace;
  let closed = false;
  try {
    const identity = await client.identify();
    opened = await client.subscribe({ treeEvents: "deltas" });
    const created = await client.newWorkspace({ name: workspaceName, cols: 80, rows: 24 });
    const surface = created.surface;
    await client.send(surface, { text: `printf '${marker}\\n'\r` });
    const waited = await client.waitFor(surface, marker, 5_000);
    const screenText = await client.readScreen(surface);
    const context = findSurface(await client.listWorkspaces(), surface);
    if (!context) throw new Error(`created surface ${surface} is absent from the tree`);
    workspace = context.workspace.id;
    const terminalCreated = context.tab.kind === "pty" && !context.tab.dead;
    await client.renameWorkspace(workspace, renamedName);
    const renamedTree = await client.listWorkspaces();
    const renamed = renamedTree.workspaces.some(
      (item) => item.id === workspace && item.name === renamedName,
    );
    await client.closeWorkspace(workspace);
    closed = true;
    const remaining = await client.listWorkspaces();
    const disappeared = remaining.workspaces.every((item) => item.id !== workspace);

    const requiredEvents = ["workspace-added", "workspace-renamed", "workspace-closed"];
    const observed = [];
    for (let index = 0; index < 64; index += 1) {
      if (requiredEvents.every((name) => observed.includes(name))) break;
      observed.push((await opened.next(Number(request.timeout_ms ?? 5_000))).event);
    }
    const positions = requiredEvents.map((name) => observed.indexOf(name));
    const streamOrdered = positions.every(
      (position, index) => position >= 0 && (index === 0 || position > positions[index - 1]),
    );
    return {
      identified: identity.protocol === 12,
      workspace_created: workspace > 0n,
      terminal_created: terminalCreated,
      marker_sent: true,
      wait_matched: waited.matched === true,
      read_contains_marker: screenText.text.includes(marker),
      stream_ordered: streamOrdered,
      renamed,
      closed,
      disappeared,
      observed_events: observed,
    };
  } finally {
    if (workspace !== undefined && !closed) {
      try {
        await client.closeWorkspace(workspace);
      } catch {
        // Preserve the primary conformance failure.
      }
    }
    opened?.close();
    await client.close();
  }
}

async function dispatch(request) {
  switch (request.op) {
    case "metadata":
      return metadata();
    case "identify":
      return identify(request);
    case "nullable-literal":
      return nullableLiteral(request);
    case "optional-non-null-response":
      return optionalNonNullResponse(request);
    case "optional-nullable-request":
      return optionalNullableRequest(request);
    case "stream":
      return stream(request);
    case "required-nullable-event":
      return requiredNullableEvent(request);
    case "optional-non-null-event":
      return optionalNonNullEvent(request);
    case "close-pending-stream":
      return closePendingStream(request);
    case "authority":
      return authority(request);
    case "authority-denied":
      return authorityDenied(request);
    case "real-flow":
      return realFlow(request);
    default:
      throw new Error(`unknown adapter operation ${String(request.op)}`);
  }
}

const chunks = [];
for await (const chunk of process.stdin) chunks.push(chunk);
const request = JSON.parse(Buffer.concat(chunks).toString("utf8").trim());
const response = { contract_version: 1, id: request.id };
try {
  response.value = await dispatch(request);
  response.ok = true;
} catch (error) {
  response.ok = false;
  response.error = { kind: classify(error), message: String(error?.message ?? error) };
}
process.stdout.write(`${JSON.stringify(response)}\n`);

File: cmux-tui/bindings/conformance/raw/adapters/python/adapter.py
#!/usr/bin/env python3
"""Protocol-10 conformance adapter for the public Python SDK."""

from __future__ import annotations

import dataclasses
import json
import sys
import threading
import time
from enum import Enum
from pathlib import Path
from typing import Any, Mapping


ROOT = Path(__file__).resolve().parents[6]
sys.path.insert(0, str(ROOT / "cmux-tui" / "bindings" / "python"))

from cmux import raw as cmux  # noqa: E402


UINT64_KEYS = {
    "client",
    "index",
    "offset",
    "pane",
    "pane_revision",
    "projection_revision",
    "request",
    "screen",
    "seq",
    "surface",
    "terminal_revision",
    "timeout_ms",
    "workspace",
    "workspace_revision",
}


def normalize(value: Any, key: str | None = None) -> Any:
    if value is cmux.MISSING:
        return None
    if isinstance(value, Enum):
        return value.value
    if dataclasses.is_dataclass(value):
        return normalize(dataclasses.asdict(value))
    if isinstance(value, Mapping):
        return {
            str(item_key): normalize(item_value, str(item_key))
            for item_key, item_value in value.items()
            if item_value is not cmux.MISSING and item_key != "raw"
        }
    if isinstance(value, (list, tuple)):
        return [normalize(item) for item in value]
    if isinstance(value, bytes):
        import base64

        return base64.b64encode(value).decode("ascii")
    if isinstance(value, int) and not isinstance(value, bool) and (
        key in UINT64_KEYS or (key is not None and key.endswith("_revision"))
    ):
        return str(value)
    return value


def event_value(event: Any) -> dict[str, Any]:
    raw = dict(getattr(event, "raw", {}) or {})
    if isinstance(event, cmux.UnknownEvent):
        return {
            "event": event.event,
            "unknown": True,
            "raw": normalize(event.raw),
        }
    if raw:
        return normalize(raw)
    return normalize(event)


def classify(error: BaseException) -> str:
    text = str(error).lower()
    if isinstance(error, cmux.TimeoutError) or "timed out" in text or "did not respond" in text:
        return "timeout"
    if "limit" in text or "exceed" in text or "too many" in text:
        return "limit"
    if isinstance(error, cmux.CommandError):
        return "command"
    if isinstance(error, cmux.ProtocolError):
        return "decode"
    return "transport"


def metadata() -> dict[str, Any]:
    commands = [
        {
            "name": item.wire_name,
            "authority": item.authority,
            "stream": item.stream_kind,
        }
        for item in cmux.COMMANDS.values()
    ]
    events = [
        {
            "name": item.wire_name,
            "streams": list(item.streams),
        }
        for item in cmux.EVENTS.values()
    ]
    return {"commands": commands, "events": events}


def make_client(
    request: Mapping[str, Any],
    *,
    allow_provider_authority: bool = False,
) -> cmux.CmuxClient:
    return cmux.CmuxClient(
        socket_path=str(request["socket_path"]),
        timeout=max(float(request.get("timeout_ms", 1000)) / 1000, 0.001),
        max_line_bytes=int(request.get("max_frame_bytes", 16 * 1024 * 1024)),
        max_pre_ack_events=int(request.get("max_buffered_events", 256)),
        allow_provider_authority=allow_provider_authority,
    )


def identify(request: Mapping[str, Any]) -> dict[str, Any]:
    with make_client(request) as client:
        value = client.identify()
        return {
            "app": value.app,
            "protocol": value.protocol,
            "workspace_revision": str(value.workspace_revision),
            "terminal_revision": str(value.terminal_revision),
        }


def nullable_literal(request: Mapping[str, Any]) -> dict[str, Any]:
    with make_client(request) as client:
        placement = client.create_terminal(key="workspace-key")
        return {"lifecycle": placement.lifecycle}


def optional_non_null_response(request: Mapping[str, Any]) -> dict[str, Any]:
    with make_client(request) as client:
        value = client.identify()
        return {"present": value.capabilities is not cmux.MISSING}


def optional_nullable_request(request: Mapping[str, Any]) -> dict[str, Any]:
    presence = str(request["presence"])
    with make_client(request) as client:
        if presence == "omitted":
            client.set_client_info()
        elif presence == "null":
            client.set_client_info(name=None)
        elif presence == "value":
            client.set_client_info(name="conformance-client")
        else:
            raise ValueError(f"unknown presence {presence!r}")
    return {"presence": presence}


def open_stream(client: cmux.CmuxClient, request: Mapping[str, Any]) -> Any:
    kind = request["stream"]
    if kind == "subscribe-coarse":
        return client.subscribe()
    if kind == "subscribe-deltas":
        return client.subscribe_deltas()
    surface = int(str(request.get("surface", "7")))
    if kind == "attach-byte":
        return client.attach_bytes(surface)
    if kind == "attach-render":
        return client.attach_render(surface)
    if kind == "attach-browser":
        return client.attach_browser(surface)
    raise ValueError(f"unknown stream {kind!r}")


def stream(request: Mapping[str, Any]) -> dict[str, Any]:
    client = make_client(request)
    events: list[dict[str, Any]] = []
    terminal = False
    try:
        opened = open_stream(client, request)
        try:
            for _ in range(int(request.get("events", 1))):
                try:
                    event = next(opened)
                except StopIteration:
                    terminal = True
                    break
                events.append(event_value(event))
                if getattr(event, "event", None) in ("overflow", "detached"):
                    terminal = True
        finally:
            opened.close()
    finally:
        client.close()
    return {"events": events, "terminal": terminal}


def required_nullable_event(request: Mapping[str, Any]) -> dict[str, Any]:
    client = make_client(request)
    opened = client.subscribe()
    try:
        event = next(opened)
        if not isinstance(event, cmux.ClientChangedEvent):
            raise cmux.ProtocolError(
                f"expected client-changed event, got {getattr(event, 'event', None)!r}"
            )
        return {"name": event.name}
    finally:
        opened.close()
        client.close()


def optional_non_null_event(request: Mapping[str, Any]) -> dict[str, Any]:
    client = make_client(request)
    opened = open_stream(client, request)
    try:
        event = next(opened)
        if not isinstance(event, cmux.OutputEvent):
            raise cmux.ProtocolError(
                f"expected output event, got {getattr(event, 'event', None)!r}"
            )
        return {"present": event.colors is not cmux.MISSING}
    finally:
        opened.close()
        client.close()


def close_pending_stream(request: Mapping[str, Any]) -> dict[str, Any]:
    client = make_client(request)
    opened = open_stream(client, request)
    finished = threading.Event()

    def read() -> None:
        try:
            next(opened)
        except BaseException:
            pass
        finally:
            finished.set()

    reader = threading.Thread(target=read, daemon=True)
    reader.start()
    time.sleep(int(request.get("close_after_ms", 50)) / 1000)
    opened.close()
    deadline = int(request.get("deadline_ms", 1000)) / 1000
    unblocked = finished.wait(deadline)
    client.close()
    reader.join(timeout=0.1)
    return {"unblocked": unblocked}


def authority(request: Mapping[str, Any]) -> dict[str, Any]:
    authority_name = str(request["authority"])
    with make_client(
        request,
        allow_provider_authority=authority_name == "provider-authority",
    ) as client:
        if authority_name == "control":
            client.ping()
            command = "ping"
        elif authority_name == "frontend":
            client.browser_back(7)
            command = "browser-back"
        elif authority_name == "local-admin":
            client.pairing_response(1, False)
            command = "pairing-response"
        elif authority_name == "provider-authority":
            client.mark_workspaces_provider_managed("conformance-authority")
            command = "mark-workspaces-provider-managed"
        else:
            raise ValueError(f"unknown authority {authority_name!r}")
    return {"command": command}


def authority_denied(request: Mapping[str, Any]) -> dict[str, Any]:
    with make_client(request) as client:
        try:
            client.mark_workspaces_provider_managed("conformance-authority")
        except cmux.AuthorityError:
            return {"denied": True}
    raise RuntimeError("default client allowed provider-authority command")


def real_flow(request: Mapping[str, Any]) -> dict[str, Any]:
    marker = str(request.get("marker", "cmux-sdk-conformance-marker"))
    workspace_name = str(request.get("workspace_name", "sdk-conformance-workspace"))
    renamed_name = str(request.get("renamed_name", "sdk-conformance-renamed"))
    client = make_client(request)
    opened = None
    workspace: int | None = None
    closed = False
    try:
        identity = client.identify()
        opened = client.subscribe_deltas()
        created = client.new_workspace(name=workspace_name, cols=80, rows=24)
        surface = created.surface
        client.send(surface, text=f"printf '{marker}\\n'\r")
        waited = client.wait_for(surface, marker, 5_000)
        screen = client.read_screen(surface)
        context = cmux.find_surface(client.list_workspaces(), surface)
        if context is None:
            raise RuntimeError(f"created surface {surface} is absent from the tree")
        workspace = context.workspace.id
        terminal_created = context.tab.kind == "pty" and not context.tab.dead
        renamed = client.rename_workspace(renamed_name, workspace=workspace)
        client.close_workspace(workspace=workspace)
        closed = True
        remaining = client.list_workspaces()
        disappeared = all(item.id != workspace for item in remaining.workspaces)

        required_events = [
            "workspace-added",
            "workspace-renamed",
            "workspace-closed",
        ]
        observed: list[str] = []
        for _ in range(64):
            if all(name in observed for name in required_events):
                break
            observed.append(next(opened).event)
        positions = [observed.index(name) for name in required_events]
        stream_ordered = positions == sorted(positions)
        return {
            "identified": identity.protocol == 12,
            "workspace_created": workspace > 0,
            "terminal_created": terminal_created,
            "marker_sent": True,
            "wait_matched": waited.matched is True,
            "read_contains_marker": marker in screen.text,
            "stream_ordered": stream_ordered,
            "renamed": renamed.workspace == workspace,
            "closed": closed,
            "disappeared": disappeared,
            "observed_events": observed,
        }
    finally:
        if workspace is not None and not closed:
            try:
                client.close_workspace(workspace=workspace)
            except BaseException:
                pass
        if opened is not None:
            opened.close()
        client.close()


def dispatch(request: Mapping[str, Any]) -> Any:
    operation = request.get("op")
    if operation == "metadata":
        return metadata()
    if operation == "identify":
        return identify(request)
    if operation == "nullable-literal":
        return nullable_literal(request)
    if operation == "optional-non-null-response":
        return optional_non_null_response(request)
    if operation == "optional-nullable-request":
        return optional_nullable_request(request)
    if operation == "stream":
        return stream(request)
    if operation == "required-nullable-event":
        return required_nullable_event(request)
    if operation == "optional-non-null-event":
        return optional_non_null_event(request)
    if operation == "close-pending-stream":
        return close_pending_stream(request)
    if operation == "authority":
        return authority(request)
    if operation == "authority-denied":
        return authority_denied(request)
    if operation == "real-flow":
        return real_flow(request)
    raise ValueError(f"unknown adapter operation {operation!r}")


def main() -> int:
    line = sys.stdin.buffer.readline()
    request: dict[str, Any] = json.loads(line)
    response: dict[str, Any] = {
        "contract_version": 1,
        "id": request.get("id"),
    }
    try:
        response["value"] = dispatch(request)
        response["ok"] = True
    except BaseException as error:
        response["ok"] = False
        response["error"] = {"kind": classify(error), "message": str(error)}
    sys.stdout.write(json.dumps(response, separators=(",", ":"), ensure_ascii=False) + "\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
